File: problems/474/474.ones-and-zeroes.py
#
# @lc app=leetcode id=474 lang=python3
#
# [474] Ones and Zeroes
#


# @lc code=start
import math
from functools import cache


class Solution:
    # def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
    #     """
    #     find subset such that
    #         at most m x "0"
    #         at most n x "1"
    #     return size of the largest subset

    #     eg:
    #     strs = ["10","0001","111001","1","0"], m = 5, n = 3

    #     dp[M][N]
    #     dp[i][j] -> largest number of items in subset for m = i, n = j
    #     """

    #     dp = [[0] * (n + 1) for _ in range(m + 1)]
    #     for s in strs:
    #         zero_cnt = s.count("0")
    #         one_cnt = len(s) - zero_cnt

    #         for i in range(m, zero_cnt - 1, -1):
    #             for j in range(n, one_cnt - 1, -1):
    #                 dp[i][j] = max(dp[i][j], dp[i - zero_cnt][j - one_cnt] + 1)
    #         print(dp)

    #     return dp[m][n]

    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        cnts = [[s.count("0"), s.count("1")] for s in strs]

        @cache
        def dp(i, j, idx):
            if i < 0 or j < 0:
                return -math.inf

            if idx == len(strs):
                return 0

            return max(
                dp(i, j, idx + 1), 1 + dp(i - cnts[idx][0], j - cnts[idx][1], idx + 1)
            )

        return dp(m, n, 0)


# @lc code=end

File: problems/474/474.ones-and-zeroes.java
/*
 * @lc app=leetcode id=474 lang=java
 *
 * [474] Ones and Zeroes
 */

// @lc code=start

import java.util.HashMap;
import java.util.Map;

class Solution {
    public int findMaxForm(String[] strs, int m, int n) {
        /*
         * return size of largest subset of strs such that at most
         * m "0"s and n "1"s
         *
         * intuition:
         * - recursive brute force dp with cache.
         * - dp(m,n)
         *
         * O(N^2) time, O(N) space.
         */

        Map<String, Integer> dp = new HashMap<>();
        return helper(strs, 0, m, n, 0, dp);
    }

    private int helper(String[] strs, int i, int m, int n, int size, Map<String, Integer> dp) {
        if (m < 0 || n < 0) {
            return size - 1;
        }

        if ((m == 0 && n == 0) || i >= strs.length) {
            return size;
        }

        String key = i + "," + m + "," + n;
        if (dp.containsKey(key)) {
            return dp.get(key);
        }

        int ones = countOnes(strs[i]);
        int zeros = strs[i].length() - ones;
        int res = Math.max(
                helper(strs, i + 1, m - zeros, n - ones, size + 1, dp),
                helper(strs, i + 1, m, n, size, dp));
        dp.put(key, res);

        return res;
    }

    private int countOnes(String str) {
        int cnt = 0;
        for (char c : str.toCharArray()) {
            if (c == '1') {
                cnt++;
            }
        }
        return cnt;
    }
}
// @lc code=end
